Fix importing of frozen modules (fixes #2322)

wnienhaus opened the pull request:
This aims to fix how frozen modules are handled/identified/distinguished.

It prefixes all frozen files with a "virtual directory", which is then added to sys.path (as opposed to before, where frozen files were found via the `''` current directory). This is essentially option #2 suggested in issue #2322.

The virtual directory is called `{frozen}`, a name chosen to be highly unlikely to ever be a real directory name. (The `:frozen:` name suggested in issue #2322 was not suitable, because the colon (:) character is used as the path separator in the string representation of sys.path)

While the initial fix for the unix port was quite easy (prefix all frozen filenames with `{frozen}/` and add `{frozen}` as the last entry in the default sys.path), it ended up being a bit of a rabbit hole, requiring more changes/fixes in the test suite and other ports, even revealing a potential bug which I will report/fix separately.

One issue I am not 100% happy about: some platforms use `pyexec_frozen_module` to run frozen py files on boot up. This function does not search paths defined in sys.path, and arguably shouldn't need to, since it uses the arrays with frozen files/strings and essentially "knows" where to find frozen files/modules. But due to the prefix of filenames in the array, it currently needs to be passed the prefix too. Changing this, so the function (and those it calls) internally skip over the prefix while matching file names is another rabbit hole, which I haven't yet gone into. So for now, `pyexec_frozen_module` must be given filenames with the prefix. I am happy for thoughts around this - maybe there is a nicer way to solve this.

Commit:
all: Move frozen modules to virtual directory.

This change "moves" frozen modules into a virtual directory
called ".frozen". This allows controlling the search order for
frozen modules, relative to other directories.

Before this change, frozen modules were found via the current
directory (i.e. the "" entry in sys.path). However, when a file
is executed from the command line the first entry in sys.path
(which is normally the empty string ""), is replaced by the base
path of the executing file (correct behaviour as per CPython).
This made loading frozen modules fail.

With this change, frozen modules have their own entry in
sys.path, which also allows the user to control the order of
search when looking for frozen modules.

By default, the virtual ".frozen" directory is the last entry
in sys.path (i.e. it gets searched last).

Fixes issue #2322.

dhalbert commented:
In CircuitPython, we use `.frozen` for this virtual directory name. We normally put libraries in `lib/`, and put `.frozen` before `lib` in `sys.path`. We use frozen modules primarily to avoid using up RAM import on boards with limited RAM, for libraries that work with on-board peripherals (like, say, an on-board accelerometer).

peterhinch commented:
> While the initial fix for the unix port was quite easy (prefix all frozen filenames with {frozen}/ and add {frozen} as the last entry in the default sys.path),

To retain existing behaviour `{frozen}` (or `.frozen`) should be the first entry in `sys.path`. By default frozen modules take precedence over matching files on the filesystem. The user can modify this by altering `sys.path`.

wnienhaus commented:
> To retain existing behaviour `{frozen}` (or `.frozen`) should be the first entry in `sys.path`. By default frozen modules take precedence over matching files on the filesystem. The user can modify this by altering `sys.path`.

You are right, on devices (I tested only esp8266/esp32) frozen modules take precedence because sys.path always has `''` as the first entry and frozen modules are searched first, however, on the unix port this is not always true: when executing a script from a file, the first entry in sys.path is replaced with the directory of the that contains the called script, and in that case frozen modules are never searched (or found). This is what this PR is mainly trying to fix. (see issue #3509 for more discussion on this).

I did not consider backwards compatibility up until now (not intentionally) but wonder if it's good that frozen modules take precedence. If frozen modules come after the current directory in search order, it would allow overriding frozen code (e.g. for testing/hot-fixing) without having to rebuild/reflash the firmware, while the other way around this is not possible. On the unix port specifically, it feels a bit more aligned with CPython (see my tests documented in #3509) and .

Would it make sense to have a different default search order on different ports? e.g. have the unix port search frozen modules last (it would still be before builtins), while bare-metal ports search them first? (Would not be my choice, but maybe it's an option?)

(Btw, I like `.frozen` so there's alignment between MicroPython forks)

dhalbert commented:
> ...If frozen modules come after the current directory in search order, it would allow overriding frozen code (e.g. for testing/hot-fixing) without having to rebuild/reflash the firmware, while the other way around this is not possible.

We use that idea to be able to override the modules we've frozen for RAM-space-saving.

CircuitPython's default `sys.path` is `['', '/', '.frozen', '/lib']`. Normally what we freeze are versions of what might be in `/lib/`. When we first added an explicit `.frozen`, we added it at the end of `sys.path`, but this did not work out so well. When we had a smaller set of libraries we actually encouraged our users to download the latest library bundle (a collection of all our libraries) and unzip the whole thing into `lib`. But that meant there were duplicate copies of libraries in `lib` that overrode the frozen ones, which negated the RAM saving-behavior of using a frozen library. This was a significant support issue for us (discussed in https://github.com/adafruit/circuitpython/issues/1310). Now there are far too many libraries to have them all in `lib`, so this is less important, but putting `.frozen` before `lib` still reduces the support queries.

To emphasize, our primary use for frozen code has been to save RAM on low-RAM boards. We don't bother on boards with more RAM. We haven't had the need yet to freeze mandatory initialization code. If we needed to do that we might consider adding another virtual directory that came earlier in `sys.path`

peterhinch commented:
> If frozen modules come after the current directory in search order, it would allow overriding frozen code (e.g. for testing/hot-fixing) without having to rebuild/reflash the firmware, while the other way around this is not possible.

It can be done by having the following in `main.py`:
```python
import sys
sys.path.append(sys.path.pop(0))
```

dpgeorge commented on tools/mpy-tool.py:
```diff
@@ -843,7 +843,7 @@ def freeze_mpy(base_qstrs, raw_codes):
     for rc in raw_codes:
         module_name = rc.source_file.str
-        print('"%s\\0"' % module_name)
+        print('"{frozen}/%s\\0"' % module_name)
```
This will increase code size of firmware by an amount proportional to the number of frozen files.  I think it would be better to have a bit of code in `py/builtinimport.c` which knows about the special frozen path.

wnienhaus replied:
Very good point. I just reviewed my change, and I think, if I make `do_load` and `mp_import_stat_any` test for the `{frozen}/` prefix and strip (skip) it before passing the path into `mp_find_frozen_module` and `mp_frozen_stat` (i.e. the functions already meant for loading frozen files will not get the `{frozen}/` prefix), then all will work, and we don't need the prefix in the frozen filenames. What do you think?

I intend to make a commit that reflects that change. But since the PR is so old, is it ok to rebase it on top of current master and force-push?

wnienhaus replied:
For now, I have made a proposed patch: https://gist.github.com/wnienhaus/4492a4128d74e19f1248deab7ed31f7c - to show in code what I mean. There's one ugly part in the patch (around the source_name for frozen strings) - see comment in the gist.

But as I was testing my change, a new idea came up: would it perhaps make sense to implement the entire frozen file handing via VFS, mounted to /.frozen/ ? This would potentially allow removing all special frozen file logic from everywhere, and we'd simply have one more directory to search like any other directory? It feels like a bigger change, but maybe it's a cleaner approach? (I do see that VFS is a compile time directive, so I guess some ports don't have VFS, but could it be that ports with frozen file support always have VFS?)

dpgeorge replied:
@wnienhaus feel free to rebase and force-push to this branch your latest implementation, that's the best approach.  Then we can re-review it in full.

Re VFS approach: this is an interesting idea, but I don't think it'll work without a lot of changes, because frozen code is nothing like a filesystem (eg there's no flat data to read for any given file, instead it's a set of nested data structures).

dpgeorge commented:
@dhalbert do you still use `.frozen` for the frozen path?  Any further updates/insights on this topic since Sept 2019?

wnienhaus commented:
@dhalbert any update on your experience with the `.frozen` path? I looked over CircuitPython code and the history of the code around `.frozen` and it looks like my most recent approach is pretty much identical.

Are you aware of any limitations/issues that could be nice to address, or does it work as expected? If so and with your permission, I would actually propose using the code as it is in CircuitPython (except perhaps removing the double check as mentioned here https://github.com/adafruit/circuitpython/pull/235#pullrequestreview-60785923).

I could update this PR accordingly.

wnienhaus commented:
I have now rebased this branch on the latest master and improved the change:
* I renamed the virtual directory from `{frozen}` to `.frozen` - to match what CircuitPython has done (and it feels easier to type in a REPL)
* I have changed to the approach where methods in `builtinimport.c` are aware of the `.frozen/` prefix, so the prefix is no longer needed in the filename encoded in the `mp_frozen_mpy_names` array.
* I have added the `.frozen` directory to `sys.path` on more ports - I tried my best to establish which ports have frozen module support enabled, but might have missed some.
* I have fixed the "extra coverage" tests, to pass with with the new `.frozen` directory behaviour

I also tested this:
* with the unix port using `make test`
* with the unix port using `make VARIANT=coverage test`
* interactively on the ESP8266, which has a number of frozen modules by default (e.g. `dht`)

wnienhaus committed:
rp2: Fix build with virtual frozen directory.

The fix ensures main.c is included when searching for qstrs like
MP_QSTR_Foo. This requires including tinyusb header files during
the qstr preprocessing step and also requires setting the
MICROPY_MODULE_FROZEN compile definition (this is for the step
that generates qstr.i.last - during the later compile stage
the define is correctly set).

wnienhaus committed:
qemu-arm: Fix build with virtual frozen directory.

wnienhaus commented:
Any update on this?

dhalbert commented:
> @dhalbert do you still use `.frozen` for the frozen path? Any further updates/insights on this topic since Sept 2019?

I apologize for not replying to this earlier. I subscribe to all MicroPython repo notifications, and this slipped by.

We still use the same path: `['', '/', '.frozen', '/lib']`. It has continued to work out well. We don't have many support issues regarding the frozen module path. There are more issues about multiple copies of library files, with some versions being out of date, such as an older .py file overriding a new .mpy. That is not a frozen module issue

jimmo commented:
I think `.frozen` makes more sense than `|frozen`, especially for CircuitPython compatibility.

A possible alternative is to use `None` or even `...` as the sentinel value? i.e. `sys.path = ['foo', ..., '/lib']` would search frozen modules after `foo`, but before `/lib`.
This is still valid in CPython too (it will ignore non-string entries). It's neat, but it would mean that you could no longer add a prefix path for frozen modules to sys.path.

For example if I'd frozen foo/bar/baz.py, the current behaviour is that I can do

```
>>> import baz
ImportError: no module named 'baz'
>>> sys.path.append('foo/bar')
>>> import baz
>>> baz
<module 'baz' from 'foo/bar/baz.py'>
```

With this PR, I'd now append `.frozen/foo/bar` instead, and TBH I think that probably makes sense. I have no idea if anyone depends on this behaviour, but I can see it being useful when freezing the entire codebase. OTOH, if we didn't want to support this, then it could simplify the import code to make it such that frozen imports are always top-level.

dpgeorge commented:
Superseded by #8079, which was merged, see 86394f70fcd7b9fa66ec952711e5d64557d81a1e

Thanks @wnienhaus for this initial implementation!
